src/lib.rs

```diff
@@ -0,0 +1,309 @@
+        pub fn add_validator(origin, v: T::AccountId) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(Self::is_validator(&who), Error::<T>::ValidatorInvalid);
+
+            ensure!(!Self::is_validator(&v), Error::<T>::ValidatorAlreadyExists);
+            <Validators<T>>::insert(&v, true);
+            Self::deposit_event(RawEvent::ValidatorAdded(v));
+            Ok(())
+        }
+
+        /// Removes an existing validator from the set. Errors if validator doesn't exist.
+        pub fn remove_validator(origin, v: T::AccountId) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(Self::is_validator(&who), Error::<T>::ValidatorInvalid);
+
+            ensure!(Self::is_validator(&v), Error::<T>::ValidatorInvalid);
+            <Validators<T>>::remove(&v);
+            Self::deposit_event(RawEvent::ValidatorRemoved(v));
+            Ok(())
+        }
+
+        pub fn create_proposal(origin, hash: T::Hash, call: Box<<T as Trait>::Proposal>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(Self::is_validator(&who), Error::<T>::ValidatorInvalid);
+
+            // Make sure proposal doesn't already exist
+            ensure!(!<Votes<T>>::contains_key(hash), Error::<T>::ProposalAlreadyExists);
+
+            let proposal = ProposalVotes::new(hash);
+            <Votes<T>>::insert(hash, proposal.clone());
+            <Proposals<T>>::insert(hash, call);
+
+            // Creating a proposal also votes for it
+            Self::vote_for(who, proposal)
+        }
+
+        pub fn vote(origin, hash: T::Hash, in_favour: bool) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(Self::is_validator(&who), Error::<T>::ValidatorInvalid);
+
+            // Check if proposal exists
+            if let Some(votes) = <Votes<T>>::get(hash) {
+                // Vote if they haven't already
+                if in_favour {
+                    Self::vote_for(who, votes)?
+                } else {
+                    Self::vote_against(who, votes)?
+                }
+            } else {
+                Err(Error::<T>::ProposalDoesNotExist)?
+            }
+
+            Ok(())
+        }
+
+        /// Completes an asset transfer to the chain by emitting an event to be acted on by the
+        /// bridge and updating the tx count for the respective chan.
+        pub fn receive_asset(origin, dest_id: Vec<u8>, to: Vec<u8>, token_id: Vec<u8>, metadata: Vec<u8>) -> DispatchResult {
+            // TODO: Limit access
+            ensure_root(origin)?;
+            // Ensure chain is whitelisted
+            if let Some(mut counter) = Chains::get(&dest_id) {
+                // Increment counter and store
+                counter.recv += 1;
+                Chains::insert(&dest_id, counter.clone());
+                Self::deposit_event(RawEvent::AssetTransfer(dest_id, counter.recv, to, token_id, metadata));
```

Comment:
Asset transfer to endowed account is missing here.

Reply:
The actual transfer will occur outside the pallet, this should be used to signal to the bridge as the last part of the transfer call. 

Working on a corresponding pallet with example implementations that use this.

Reply:
Cool – just for my understanding: What's the reason behind having this outside of the pallet?

Reply:
This is aimed to make the bridge pallet modular. By keeping the actual transfer opaque this could be called from a variety of transfer methods without modification.